crates/stoffel-vm/src/net/discovery.rs

```diff
@@ -0,0 +1,175 @@
+                                    let peers: Vec<(PartyId, SocketAddr)> = st.iter()
+                                        .filter(|(pid, _)| **pid != party_id)
+                                        .map(|(pid, addr)| (*pid, *addr))
+                                        .collect();
+                                    let reply = DiscoveryMessage::PeerList { peers };
+                                    let _ = send_ctrl(&*conn, &reply).await;
+
+                                    // notify others if new
+                                    if is_new {
+                                        let joined = DiscoveryMessage::PeerJoined { party_id, listen_addr };
+                                        let _ = send_ctrl(&*conn, &joined).await;
+                                    }
+                                }
+                                DiscoveryMessage::RequestPeers => {
+                                    let st = state.lock().await;
+                                    let peers: Vec<(PartyId, SocketAddr)> = st.iter()
+                                        .map(|(pid, addr)| (*pid, *addr)).collect();
+                                    let _ = send_ctrl(&*conn, &DiscoveryMessage::PeerList { peers }).await;
+                                }
+                                DiscoveryMessage::Heartbeat => {
+                                    // Ignored in this minimal version
+                                }
+                                _ => {}
+                            }
+                        } else if let Ok(ps) = bincode::deserialize::<ProgramSyncMessage>(&buf) {
+                            // Handle program sync messages
+                            match ps {
+                                ProgramSyncMessage::ProgramAnnounce { .. } => {
+                                    // Respond with program sync logic if needed
+                                    // For now, just echo back
+                                    let _ = send_prog_ctrl(&*conn, &ps).await;
+                                }
+                                ProgramSyncMessage::ProgramFetchRequest { program_id } => {
+                                    // Send cached program bytes if available
+                                    let bytes_opt = uploader_bytes.lock().await.clone();
+                                    if let Some(bytes) = bytes_opt {
+                                        let _ = send_program_bytes(&*conn, program_id, bytes).await;
+                                    }
+                                }
+                                _ => {}
+                            }
+                        } else if let Ok(sess) = bincode::deserialize::<SessionMessage>(&buf) {
+                            match sess {
+                                SessionMessage::SessionAnnounce(_) => { /* not expected from clients */ }
+                                SessionMessage::SessionAck { .. } => { /* optional: track */ }
```

Comment:
Is this necessary? Wouldn't Quic be enough here?

Reply:
Ahh, that's makes sense. We'd need to flesh this out a bit though. Add a TODO and a linear issue so we can keep track of this.